src/System.Net.Security/src/System/Net/Security/SniHelper.cs

```diff
@@ -0,0 +1,278 @@
+        //   - opaque extension_data
+        private static string GetSniFromExtension(ReadOnlySpan<byte> extension, out ReadOnlySpan<byte> remainingBytes)
+        {
+            if (extension.Length < 2)
+            {
+                remainingBytes = ReadOnlySpan<byte>.Empty;
+                return null;
+            }
+
+            int extensionType = ReadUint16(extension);
+            ReadOnlySpan<byte> extensionData = extension.Slice(2);
+
+            if (extensionType == 0x00)
+            {
+                return GetSniFromServerNameList(extensionData, out remainingBytes);
+            }
+            else
+            {
+                remainingBytes = SkipOpaqueType2(extensionData);
+                return null;
+            }
+        }
+
+        // 3.1. https://www.ietf.org/rfc/rfc3546.txt
+        // ServerNameList structure:
+        //   - ServerName server_name_list<1..2^16-1>
+        // ServerName structure:
+        //   - NameType name_type (1 byte) => 0x00 is host_name
+        //   - opaque HostName
+        // Per spec:
+        //   If the hostname labels contain only US-ASCII characters, then the
+        //   client MUST ensure that labels are separated only by the byte 0x2E,
+        //   representing the dot character U+002E (requirement 1 in section 3.1
+        //   of [IDNA] notwithstanding). If the server needs to match the HostName
+        //   against names that contain non-US-ASCII characters, it MUST perform
+        //   the conversion operation described in section 4 of [IDNA], treating
+        //   the HostName as a "query string" (i.e. the AllowUnassigned flag MUST
+        //   be set). Note that IDNA allows labels to be separated by any of the
+        //   Unicode characters U+002E, U+3002, U+FF0E, and U+FF61, therefore
+        //   servers MUST accept any of these characters as a label separator.  If
+        //   the server only needs to match the HostName against names containing
+        //   exclusively ASCII characters, it MUST compare ASCII names case-
+        //   insensitively.
+        private static string GetSniFromServerNameList(ReadOnlySpan<byte> serverNameListExtension, out ReadOnlySpan<byte> remainingBytes)
+        {
+            if (serverNameListExtension.Length < 2)
+            {
+                remainingBytes = ReadOnlySpan<byte>.Empty;
+                return null;
+            }
+
+            int serverNameListLength = ReadUint16(serverNameListExtension);
+            ReadOnlySpan<byte> serverNameList = serverNameListExtension.Slice(2);
+
+            if (serverNameListLength > serverNameList.Length)
+            {
+                remainingBytes = ReadOnlySpan<byte>.Empty;
+                return null;
+            }
+
+            remainingBytes = serverNameList.Slice(serverNameListLength);
+            ReadOnlySpan<byte> serverName = serverNameList.Slice(0, serverNameListLength);
+
+            if (serverName.Length < 3)
```

Comment:
It would be clearer if the ServerName structure were documented right here, to know "why 3?"